include/os/freebsd/zfs/sys/zfs_vfsops_os.h

```diff
@@ -46,10 +50,16 @@
 extern "C" {
 #endif
 
-#ifdef TEARDOWN_INACTIVE_RMS
+#ifdef TEARDOWN_RMS
 typedef struct rmslock zfs_teardown_lock_t;
 #else
-#define	zfs_teardown_lock_t krwlock_t
+#define	zfs_teardown_lock_t		rrmlock_t
+#endif
+
+#ifdef TEARDOWN_INACTIVE_RMS
+typedef struct rmslock zfs_teardown_inactive_lock_t;
+#else
+#define	zfs_teardown_inactive_lock_t krwlock_t
```

Comment:
One question, could we instead tackle this issue by mapping the existing `rrw_*` / `rrm_*` interfaces to the right FreeBSD locking primitive in the same way the other locks types are handled (mutexes, rwlocks)?  Or are the semantics and interfaces just too different to accommodate this.

Reply:
rrm feature-wise it supports tracking who has the read lock so you can assert on it if you want and allows certain read-locks attempts to progress despite pending write waiters, neither of which are implemented in rms (it can be done but I have not seen any need to do it). Apart from that rrm, despite its performance woes, offers a different tradeoff for write locking -- take up to n locks and be done with it, while rms sends an IPI to all CPUs. Tolerable for the time being for unmount or other relatively rare events, but may not be ok for whatever upcoming rms use may show up.

tl;dr too different :)